In [3]:
import pandas as pd

In [4]:
dados = pd.read_csv("data/Customer-Churn.csv")

In [5]:
dados.shape

(7043, 18)

In [6]:
dados.head()

Maior65Anos Conjuge Dependentes  MesesDeContrato TelefoneFixo  \
0            0     Sim         Nao                1          Nao   
1            0     Nao         Nao               34          Sim   
2            0     Nao         Nao                2          Sim   
3            0     Nao         Nao               45          Nao   
4            0     Nao         Nao                2          Sim   

  VariasLinhasTelefonicas ServicoDeInternet SegurancaOnline BackupOnline  \
0    SemServicoTelefonico               DSL             Nao          Sim   
1                     Nao               DSL             Sim          Nao   
2                     Nao               DSL             Sim          Sim   
3    SemServicoTelefonico               DSL             Sim          Nao   
4                     Nao       FibraOptica             Nao          Nao   

  SeguroNoDispositivo SuporteTecnico TVaCabo StreamingDeFilmes TipoDeContrato  \
0                 Nao            Nao     Nao               Nao    Mensalmente   
1                 Sim            Nao     Nao               Nao          UmAno   
2                 Nao            Nao     Nao               Nao    Mensalmente   
3                 Sim            Sim     Nao               Nao          UmAno   
4                 Nao            Nao     Nao               Nao    Mensalmente   

  PagamentoOnline FormaDePagamento  ContaMensal Churn  
0             Sim    ChequeDigital        29.85   Nao  
1             Nao      ChequePapel        56.95   Nao  
2             Sim      ChequePapel        53.85   Sim  
3             Nao    DebitoEmConta        42.30   Nao  
4             Sim    ChequeDigital        70.70   Sim

In [7]:
# modificacao de forma manual
traducao_dict = {
    "Sim" : 1,
    "Nao" : 0
}

dados_modificados = dados[["Conjuge", "Dependentes", "TelefoneFixo", "PagamentoOnline", "Churn"]].replace(traducao_dict)
dados_modificados.head()

Conjuge  Dependentes  TelefoneFixo  PagamentoOnline  Churn
0        1            0             0                1      0
1        0            0             1                0      0
2        0            0             1                1      1
3        0            0             0                0      0
4        0            0             1                1      1

In [8]:
# modificacao pelo get_dummies
dummy_dados = pd.get_dummies(dados.drop(["Conjuge", "Dependentes", "TelefoneFixo", "PagamentoOnline", "Churn"], axis=1))

# juncao dos dados transformados com os que ja tinhamos
dados_final = pd.concat([dados_modificados, dummy_dados], axis=1)
dados_final.head()

Conjuge  Dependentes  TelefoneFixo  PagamentoOnline  Churn  Maior65Anos  \
0        1            0             0                1      0            0   
1        0            0             1                0      0            0   
2        0            0             1                1      1            0   
3        0            0             0                0      0            0   
4        0            0             1                1      1            0   

   MesesDeContrato  ContaMensal  VariasLinhasTelefonicas_Nao  \
0                1        29.85                            0   
1               34        56.95                            1   
2                2        53.85                            1   
3               45        42.30                            0   
4                2        70.70                            1   

   VariasLinhasTelefonicas_SemServicoTelefonico  ...  StreamingDeFilmes_Nao  \
0                                             1  ...                      1   
1                                             0  ...                      1   
2                                             0  ...                      1   
3                                             1  ...                      1   
4                                             0  ...                      1   

   StreamingDeFilmes_SemServicoDeInternet  StreamingDeFilmes_Sim  \
0                                       0                      0   
1                                       0                      0   
2                                       0                      0   
3                                       0                      0   
4                                       0                      0   

   TipoDeContrato_DoisAnos  TipoDeContrato_Mensalmente  TipoDeContrato_UmAno  \
0                        0                           1                     0   
1                        0                           0                     1   
2                        0                           1                     0   
3                        0                           0                     1   
4                        0                           1                     0   

   FormaDePagamento_CartaoDeCredito  FormaDePagamento_ChequeDigital  \
0                                 0                               1   
1                                 0                               0   
2                                 0                               0   
3                                 0                               0   
4                                 0                               1   

   FormaDePagamento_ChequePapel  FormaDePagamento_DebitoEmConta  
0                             0                               0  
1                             1                               0  
2                             1                               0  
3                             0                               1  
4                             0                               0  

[5 rows x 39 columns]

In [9]:
dados_final.shape

(7043, 39)

---
# Para saber mais: divisão dos dados em treino e teste

Separar o conjunto de dados em **treino** e **teste** é uma parte imortante para conseguirmos medir o desempenho real de um modelo de machine learning. Essa etapa consiste em dedicar parte do conjunto de dados para treinar o modelo e uma outra parte para teste. Mas qual o motivo de fazer isso?

Imagine que você vai fazer uma prova e o(a) professor(a) passa uma atividade de revisão antes para treinar em casa. Chegando na prova, você percebe que as questões são exatamente iguais à revisão passada pelo(a) professor(a). Como você já sabia as questões, tirou 10! Isso é exatamente o que acontece quando utilizamos os mesmos dados para treinar e testar.

Como os dados são conhecidos pelo modelo, ele simplesmente vai acertar tudo (ou quase tudo) na etapa de teste, pois são os mesmos dados. Para resolver isso, dividimos o conjunto de dados em treino e teste, sendo esse último para simular dados nunca antes vistos pelo modelo.

## Dados de treino

Os dados de treino são aqueles utilizados para a criação e treinamento do modelo. Normalmente a maioria dos dados, cerca de 70%, são utilizados para treinamento.

## Dados de teste

Os dados de teste são utilizados para comprovar que o modelo realmente funciona. Eles não são utilizados no treinamento do modelo e normalmente representam 30% da totalidade dos dados.

Além disso, no momento de reailzar a separação desses dados é importante que ela seja feita de forma aleatória, para garantirmos que não haverá nenhum padrão no momento de divisão dos dados. Assim, cada amostra terá a mesma probabilidade de ser selecionada.

---

# KNN por trás dos panos

In [10]:
xmaria = [[0,0,1,1,0,0,39.90,1,0,0,0,1,0,1,0,0,0,0,1,1,1,0,0,1,0,1,0,0,0,0,1,0,0,1,0,0,0,1]]

In [11]:
#Ymaria = ?

In [12]:
# Divisão em inputs e outputs
x = dados_final.drop("Churn", axis=1)
y = dados_final["Churn"]

In [13]:
# Biblioteca para padronizar os dados
from sklearn.preprocessing import StandardScaler

In [14]:
norm = StandardScaler()
x_normalizado = norm.fit_transform(x)
x_normalizado

array([[ 1.03453023, -0.65401193, -3.05401039, ...,  1.40641839,
        -0.54480692, -0.5298852 ],
       [-0.96662231, -0.65401193,  0.32743831, ..., -0.71102597,
         1.83551265, -0.5298852 ],
       [-0.96662231, -0.65401193,  0.32743831, ..., -0.71102597,
         1.83551265, -0.5298852 ],
       ...,
       [ 1.03453023,  1.5290241 , -3.05401039, ...,  1.40641839,
        -0.54480692, -0.5298852 ],
       [ 1.03453023, -0.65401193,  0.32743831, ..., -0.71102597,
         1.83551265, -0.5298852 ],
       [-0.96662231, -0.65401193,  0.32743831, ..., -0.71102597,
        -0.54480692,  1.88720123]])

In [15]:
x_normalizado[0]

array([ 1.03453023, -0.65401193, -3.05401039,  0.8297975 , -0.43991649,
       -1.27744458, -1.16032292, -0.96332984,  3.05401039, -0.85417615,
        1.38171215, -0.88565976, -0.52592737,  1.00669571, -0.52592737,
       -0.63393298, -0.88361972, -0.52592737,  1.3782406 ,  1.12942721,
       -0.52592737, -0.72396752,  1.0138687 , -0.52592737, -0.63943864,
        1.2273572 , -0.52592737, -0.79013171,  1.23648892, -0.52592737,
       -0.79607014, -0.56297505,  0.90418382, -0.51424938, -0.52504733,
        1.40641839, -0.54480692, -0.5298852 ])

In [16]:
xmaria_normalizado = norm.transform(pd.DataFrame(xmaria, columns=x.columns))
xmaria_normalizado

array([[-0.96662231, -0.65401193,  0.32743831,  0.8297975 , -0.43991649,
        -1.31816495, -0.82630172,  1.03806605, -0.32743831, -0.85417615,
        -0.72373975,  1.12910177, -0.52592737,  1.00669571, -0.52592737,
        -0.63393298, -0.88361972, -0.52592737,  1.3782406 ,  1.12942721,
         1.90140322, -0.72396752, -0.98632101,  1.90140322, -0.63943864,
         1.2273572 , -0.52592737, -0.79013171, -0.80874158, -0.52592737,
         1.25617071, -0.56297505, -1.1059698 ,  1.94458183, -0.52504733,
        -0.71102597, -0.54480692,  1.88720123]])

---
## Distância Euclidiana:
$$\LARGE\sqrt{\sum_{i=1}^k (a_i - b_i)^2}$$

---

In [17]:
import numpy as np

In [18]:
a = xmaria_normalizado
type(a)

numpy.ndarray

In [19]:
b = x_normalizado[0]
type(b)

numpy.ndarray

In [20]:
# 1 - começamos subtraindo
a - b

array([[-2.00115254,  0.        ,  3.3814487 ,  0.        ,  0.        ,
        -0.04072036,  0.3340212 ,  2.00139589, -3.3814487 ,  0.        ,
        -2.10545189,  2.01476153,  0.        ,  0.        ,  0.        ,
         0.        ,  0.        ,  0.        ,  0.        ,  0.        ,
         2.42733059,  0.        , -2.00018971,  2.42733059,  0.        ,
         0.        ,  0.        ,  0.        , -2.0452305 ,  0.        ,
         2.05224086,  0.        , -2.01015362,  2.45883121,  0.        ,
        -2.11744436,  0.        ,  2.41708643]])

In [21]:
# 2 - depois realizamos a exponenciação
np.square(a-b)

array([[4.00461149e+00, 0.00000000e+00, 1.14341953e+01, 0.00000000e+00,
        0.00000000e+00, 1.65814802e-03, 1.11570160e-01, 4.00558550e+00,
        1.14341953e+01, 0.00000000e+00, 4.43292768e+00, 4.05926401e+00,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
        5.89193381e+00, 0.00000000e+00, 4.00075888e+00, 5.89193381e+00,
        0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
        4.18296779e+00, 0.00000000e+00, 4.21169253e+00, 0.00000000e+00,
        4.04071758e+00, 6.04585093e+00, 0.00000000e+00, 4.48357062e+00,
        0.00000000e+00, 5.84230682e+00]])

In [22]:
# 3 - a soma
np.sum(np.square(a-b))

84.07574038273466

In [23]:
# 4 - então tiramos a raiz e temos nossa distância
np.sqrt(np.sum(np.square(a-b)))

9.169282435541762

---
# Para saber mais: funções de distância

No algoritmo KNN, para que seja feita a classificação dos registros com base nos vizinhos mais próximos, é necessário utilizar alguma medida para identificar o quão próximo o registro está dos seus vizinhos. Existem diversas medidas para se avaliar essa proximidade entre variáveis numéricas que recebem o nome de medidas de distância. Quanto maios o valor dessa medida, mais distante um elemento está do outro, ou seja, menor a similaridade entre eles.

Considere a tabela a seguir com duas observações e n variáveis para o entendimento das medidas de distância.

| Observação | Variável 1 | Variável 2 | ... | Variável n |
| --- | --- | --- | --- | --- |
| X | X1 | X2 | ... | Xn |
| Y | Y1 | Y2 | ... | Yn |

* ### Distância Euclidiana

A medida de distância mais conhecida e usada é a **distância euclidiana**. Ela consiste em subtrair as coordenadas de uma observação pela outra observação, elevar ao quadrado os resultados, somar todos os valores e extrair a raiz quadrada.

$$\Large d = \sqrt{(X_1 - Y_1)^2 + ... + (X_n - Y_n)^2}$$

* ### Distância de Manhattan

Uma distância que considera apenas a soma dos módulos das diferenças entre cada par de coordenadas.

$$\Large d = |X_1 - Y_1| + ... + |X_n - Y_n|$$

* ### Distância de Minkowski

Uma medida de distância que é a generalização de outras distâncias, como a distância euclidiana e a de Manhattan. Consiste em extrair o módulo da diferença entre cada par de coordenadas elevando o resultado a m, realizar a soma de todos os termos e, por fim, tirar a raiz m-ésima, em que m é um número qualquer. A distância euclidiana é um caso à parte quando m é igual a 2 e a distância de Manhattan é um caso à parte quando m é igual a 1.

$$\Large d = \sqrt[m]{|X_1 - Y_1|^m + ... + |X_n - Y_n|^m}$$

* ### Distância de Chebyshev

É uma medida de distância que considera apenas o valor máximo entre os módulos das diferenças entre as variáveis. Dessa forma, leva em consideração apenas a variável que possui a maior diferença de valores entre as duas observações.

$$\Large d = máx{|X_1 - Y_1|, ..., |X_n - Y_n|}$$

Existem diversas outras medidas de distância que podem ser calculadas e cada uma delas impacta diretamente no resultado do modelo. As distâncias apresentadas aqui podem ser utilizadas no algoritmo KNN do scikit-learn, sendo a distância de Minkowski a medida de distância padrão ao instanciar o algoritmo, como pode ser checado na [documentação](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html).

Podemos modificar a medida de distância utilizando o argumento `metric` da função `sklearn.neighbors.KNeighborsClassifier`. Com base nas medidas que foram apresentadas no texto, o parâmetro pode receber os seguintes valores:

    * "euclidean" para a distância euclidiana;
    * "manhattan" para a distância de Manhattan;
    * "minkowski" para a distância de Minkowski;
    * "chebyshev" para a distância de Chebyshev.

Caso tenha interesse em descobrir o funcionamento ou outras medidas de distância que podem ser utilizadas, confira a documentação do [sklearn.metrics.DistanceMetric](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.DistanceMetric.html#sklearn.metrics.DistanceMetric).

---

### Implementando o modelo

In [24]:
# Biblioteca para divisão dos dados
from sklearn.model_selection import train_test_split

In [25]:
x_treino, x_teste, y_treino, y_teste = train_test_split(x_normalizado, y,
                                                        test_size=0.3, random_state=123)

### Treino e teste  

In [26]:
# Biblioteca para criarmos o mmodelo de machine learning
from sklearn.neighbors import KNeighborsClassifier

In [27]:
# Instanciar o modelo (criamos o modelo) - por padrão são 5 vizinhos
knn = KNeighborsClassifier(metric="euclidean")

In [28]:
# Treinando o modelo com os dados de treino
knn.fit(x_treino, y_treino)

KNeighborsClassifier(metric='euclidean')

In [29]:
# Testando o modelo com os dados de teste
predito_knn = knn.predict(x_teste)
predito_knn

array([1, 0, 0, ..., 0, 0, 0])

## KNN mão na massa

In [30]:
import numpy as np

def calcula_distancia(dados_clientes, dados_maria, numero_clientes):
    distancias = []

    # loop para percorrer cliente por cliente
    for i in range(numero_clientes):
        dist1 = dados_maria - dados_clientes[i] # subtração dos atributos
        soma_quadrado = np.sum(np.square(dist1)) # exponenciação e soma dos atributos
        distancias.append(np.sqrt(soma_quadrado)) # raiz quadrada da soma anterior

    return distancias

calcula_distancia(x_normalizado, xmaria_normalizado, 10)

[9.169282435541762,
 8.742314068192332,
 7.96104771727276,
 9.683687116435184,
 7.35756725091328,
 8.58650261645753,
 8.411628863624525,
 10.145869884558158,
 8.97159048640628,
 7.5587041367106105]

---
### Fórmula do teorema de Naïve Bayes:

$$\Large P(y|X) = \dfrac{P(X|y) * P(y)}{P(X)}$$

---
### Multivariada de Bernoulli:

$$\Large P(x_i|y) = P(i|y)x_i + (1 - P(i|y))(1 - x_i)$$

#### Os dados estão em variáveis binárias.
---

### Para saber mais: probabilidade condicional e Bayes

O teorema de Bayes é uma fórmula utilizada para calcular a probabilidade de um evento ocorrer sabendo que um outro evento, chamado de condicionante, já ocorreu, denominado **probabilidade condicional**. Sua notação é dada por P(A|B), que significa a probabilidade de A dado que B já aconteceu e é definido pela seguinte equação:

$$\Large P(A|B) = \dfrac{P(B|A) * P(A)}{P(B)}$$

Onde:

    * P(B|A): probabilidade de B acontecer dado que A já aconteceu;
    * P(A): probabilidade de A acontecer;
    * P(B): probabilidade de B acontecer.

Falando dessa forma, pode ficar complicado de visualizar na prática o que isso realmente significa. Como o teorema de Bayes envolve probabilidade, pode ser aplicado a inúmeros contextos. Vamos utilizar aqui um exemplo envolvendo a área da saúde, mas o teorema poderia ser utilizado em outras áreas. Considere o exemplo:

| | Covid-19 (5%) | Sem Covid-19 (95%) |
| --- | --- | --- |
| Teste positivo | 85% | 10% |
| Teste negativo | 15% | 90% |

Primeiramente, vamos entender o que está na tabela. Sabemos, através de pesquisas realizadas anteriormente, que há uma probabilidade de 5% de ter Covid-19. Consequentemente, há uma probabilidade de 95% de não ter. Quem tem a doença está na coluna "Covid-19 (5%)" e possui uma probabilidade de 85% de testar positivo e 15% de negativo. Já quem não tem está na coluna "Sem Covid-19 (95%)" e possui uma possibilidade de 10% de testar positivo e 90% de negativo.

De posse dessas informações, vamos repsonder à pergunta:
quais as chances de ter Covid-19 dado que o teste deu positivo?

    * Evento A: ter Covid-19
    * Evento B positivo no teste (evento condicionante)

Utilizando a fórmula apresentada anteriormente, precisamos definir algumas probabilidades, como:

    * P(B|A): probabilidade do teste ser positivo dado que a pessoa tem Covid-19, que é de 85% ou 0.85, de acordo com a tabela.
    * P(A): probabilidade de ter Covid-19. Observando a tabela, temos que é de 5%, ou 0.05.
    * P(B): a probabilidade do teste ser positivo.

A P(B) não conseguimos encontrar diretamente na tabela, pois a probabilidade do teste ser positivo pode acontecer quando a pessoa testada possui a doença ou não (chamado também de complementar). Logo, a probabilidade P(B) pode ser calculada como:

$$\Large P(B) = P(B|A) * P(A) + P(B|A^c) * P(A^c)$$

    * P(B|AC): probabilidade do teste ser positivo dado que a pessoa não possui Covid-19, que é 10% ou 0.1.
    * P(AC): probabilidade da pessoa não ter Covid-19, que é 95% ou 0.95.

Novamente, as informções acima foram retiradas diretamente da tabela. Pronto, agora temos todas as informações que precisamos. Colocando na equação temos que:

$$\Large P(A|B) = \dfrac{P(B|A) * P(A)}{P(B)} = \dfrac{0.85 * 0.05}{0.1375} = 0.31$$

Logo, as chances de ter Covid-19, dado que o teste deu positivo, são de 31%.

---

# Treinamento e teste

In [31]:
x_treino

array([[-0.96662231, -0.65401193, -3.05401039, ...,  1.40641839,
        -0.54480692, -0.5298852 ],
       [-0.96662231,  1.5290241 ,  0.32743831, ...,  1.40641839,
        -0.54480692, -0.5298852 ],
       [-0.96662231, -0.65401193,  0.32743831, ..., -0.71102597,
         1.83551265, -0.5298852 ],
       ...,
       [ 1.03453023,  1.5290241 ,  0.32743831, ...,  1.40641839,
        -0.54480692, -0.5298852 ],
       [ 1.03453023,  1.5290241 , -3.05401039, ..., -0.71102597,
         1.83551265, -0.5298852 ],
       [-0.96662231, -0.65401193,  0.32743831, ...,  1.40641839,
        -0.54480692, -0.5298852 ]])

In [32]:
y_treino

1479    0
2377    0
6613    0
6468    0
2668    1
       ..
5218    0
4060    0
1346    1
3454    0
3582    0
Name: Churn, Length: 4930, dtype: int64

In [33]:
# Escolho utilizar mediana, porque é o valor central dos nossos dados ordenados
np.median(x_treino)

-0.5259273725834717

In [34]:
# Biblioteca para criamos o modelo de machine learning
from sklearn.naive_bayes import BernoulliNB

In [35]:
# Criamos o modelo
bnb = BernoulliNB(binarize=-0.5259273725834717)

In [36]:
bnb.fit(x_treino, y_treino)

BernoulliNB(binarize=-0.5259273725834717)

In [37]:
predito_bnb = bnb.predict(x_teste)
predito_bnb

array([1, 0, 0, ..., 0, 1, 0])

# Árvore de Decisão ou Decision Tree

### Critérios de divisão de nós:

    * Índice Gini
    * Qui-Quadrado
    * Ganho de informação (entropia)
    * Redução na variância

---
## Índice Gini:

$$\Large G(node) = \sum_{k=1}^c p_k(1 - p_k)$$

$p_k$ = Probabilidade de ocorrência de um dado da classe k

$(1 - p_k)$ = Probabilidade de não ocorrência de um dado da classe k

---
## Entropia:

$$\Large Entropy(node) = -\sum_{i=1}^c p_k log(p_k)$$

---

## Para saber mais: como funciona a árvore de decisão?

Se você está se sentindo doente, mas não faz ideia do que possa ser, quando você vai ao médico são feitas uma série de perguntas (dor de cabeça? Febre? Tosse?...). Ao final da série de perguntas realizadas pelo médioco é emitido um parecer informando qual a causa dos sintomas. Nessa situação apresentada foi seguido o mesmo princípio utililzado em um estimador chamado **árvore de decisão**.

Árvore de decisão é um dos modelos de previsão mais simples, inspirada na forma que os seres humanos tomam as decisões e tem uma alta interpretabilidade, ou seja, uma compreensão fácil dos passos que foram realizados para conseguir chegar ao resultado final. Ela pode ser utilizada tanto para modelos de regressão, que têm intuito de prever valores numéricos, quanto para modelos de classificação, que têm intuito de prever categorias.

Graficamente, a árvore de decisão pode ser representada de forma que cada uma das decisões tomadas no processo possam ser visualizadas. Seus elementos principais são os **nós**, **ramos** e **folhas**. A estrutura da árvore se inicia com um **nó inicial**, também chamado de **raiz**. A partir dela são traçadas **ramificações**, que geram novos nós e o processo se repete para os nós subsequentes até que chegue a uma **folha**, que se trata de um nó especial que tem a informação da resposta, sendo ela uma categoria ou um valor previsto.

Cada ramo representa uma tomada de decisão a partir de um valor ou de uma categoria das variáveis explicativas, dividindo o conjunto de dados em nós que apresentam dados com características cada vez mais similares entre si.

Para que fiquem mais compreensíveis esses conceitos, vamos a um exemplo.

![Orientações sobra a Covid-19](https://caelum-online-public.s3.amazonaws.com/2422-machine-learning/04/Aula4-img1.png)

Na imagem de exemplo, temos umma orientação sobre Covid-19 para que as pessoas sejam direcionadas a ficarem isoladas ou não. A raiz ou nó inicial representa a pergunta se a pessoa está com sintomas de Covid-19, como aquelas perguntas realizadas pelo médico indicadas no início desse texto. A pergunta é respondida através dos ramos que partem da raiz, separando as pessoas qeu possuem sintomas das pessoas que não possuem sintomas.

O nó referente às pessoas que possuem sintomas se trata de um nó folha, com a decisão final de isolamento social. O nó referente às pessoas que não possuem sintomas se trata de um nó interno, que passa por um novo questionamento, criando assim novos ramos e nós. O processo se repete até que se chegue em decisões finais.

Portanto, o esquema se trata de uma Árvore de Decisão, em que é possível detectar todas as escolhas que foram feitas para se chegar às conclusões finais. O algoritmo de computaodr seguirá esses mesmos princípios, tomando as decisões com base nas variáveis explicativas.

### Critério de divisão dos nós

Para conseguir identificar qual o melhor momento em que um nó deve ser dividido em dois ou mais subnós, o algoritmo da árvore de decisão considera alguns critérios. Os dois principais critérios de divisão usados nas árvores de decisão são:

#### Índice Gini

Este índice informa o **grau de heterogeneidade** dos dados. O objetivo dele é medir a frequência de um elemento aleatório de um nó ser rotulado de maneira incorreta. Em outros termos, esse índice é capaz de medir a impureza de um nó e ele é determinado por meio do seguinte cálculo:

$$\Large Gini = 1 - \sum_{i=1}^k P(i)^2$$

Onde:

* $P(i)$ representa a frequência relativa das classes em cada um dos nós;

* $k$ é o número de classes.

Se o índice Gini for igual a 0, isso significa que o nó é puro. No entanto, se o valor dele se aproxima mais do valor 1, o nó é impuro.

#### Entropia

A ideia básica da entropia é medir a **desordem dos dados** de um nó por meio da variável classificadora. Assim como o índice Gini, ela é utilizada para caracterizar a impureza dos dados e pode ser calculada por meio da seguinte fórmula:

$$\Large Entropia = \sum_{i=1}^c -p_i * log_2(p_i)$$

Onde:

*  $p_i$ representa a proporção de dados no conjunto de dados (S), pertencentes à classe específica $i$;

* $c$ é o número de classes.

In [38]:
# Biblioteca para criarmos o modelo de machine learning
from sklearn.tree import DecisionTreeClassifier

In [39]:
# Instanciando o modelo
dtc = DecisionTreeClassifier(criterion="entropy", random_state=42)

In [40]:
# Treinar o modelo
dtc.fit(x_treino, y_treino)

DecisionTreeClassifier(criterion='entropy', random_state=42)

In [41]:
# Verificar a importância de cada atributo
dtc.feature_importances_

array([0.03168742, 0.01862901, 0.00064085, 0.02675693, 0.0208788 ,
       0.2056003 , 0.29546501, 0.0061786 , 0.00273035, 0.00747536,
       0.        , 0.04908356, 0.        , 0.01609395, 0.        ,
       0.00722833, 0.01220706, 0.        , 0.01248286, 0.01138295,
       0.0006673 , 0.0095669 , 0.01841631, 0.        , 0.00650511,
       0.00385849, 0.        , 0.00666929, 0.00517146, 0.        ,
       0.00665371, 0.00473251, 0.15803541, 0.00348686, 0.01414378,
       0.01601964, 0.01064163, 0.01091026])

In [42]:
predito_dtc = dtc.predict(x_teste)
predito_dtc

array([1, 0, 0, ..., 1, 0, 0])

## Validação dos modelos

### Matriz de confusão

![Matriz de Confusão](https://diegonogare.net/wp-content/uploads/2020/04/matrizConfusao-600x381.png)

In [43]:
from sklearn.metrics import confusion_matrix

In [44]:
print(confusion_matrix(y_teste, predito_knn))

[[1302  221]
 [ 317  273]]


In [45]:
print(confusion_matrix(y_teste, predito_bnb))

[[1051  472]
 [ 116  474]]


In [46]:
print(confusion_matrix(y_teste, predito_dtc))

[[1249  274]
 [ 290  300]]


## Acurácia ou *Accuracy*

A partir do cálculo da matriz de confusão conseguimos inferir outras métricas, como por exemplo a acurácia.

$$\Large ACC = \dfrac{TP + TN}{TP + FP + TN + FN}$$

In [47]:
from sklearn.metrics import accuracy_score

In [48]:
# Modelo KNN
print(accuracy_score(y_teste, predito_knn))

0.7453857075248462


In [49]:
# Modelo Bernoulli de Naïve Bayes
print(accuracy_score(y_teste, predito_bnb))

0.7217226691907241


In [50]:
# Modelo Árvore de Decisão
print(accuracy_score(y_teste, predito_dtc))

0.7330809275911027


## Precisão ou *Precision*

Outra métrica importante é a precisão, que calcula quantos foram classificados corretamente como positivos (**TP**).

$$\Large PS = \dfrac{TP}{TP + FP}$$

In [51]:
from sklearn.metrics import precision_score

In [52]:
# Modelo KNN
print(precision_score(y_teste, predito_knn))

0.5526315789473685


In [53]:
# Modelo Bernoulli de Naïve Bayes
print(precision_score(y_teste, predito_bnb))

0.5010570824524313


In [54]:
# Modelo Árvor de Decisão
print(precision_score(y_teste, predito_dtc))

0.5226480836236934


## Recall

Outra métrica importante é a *Recall* ou revocação ou ainda sensibilidade, que calcula o quão bom o modelo está para classificar corretamente um resultado positivo (**TP**).

$$\Large RC = \dfrac{TP}{TP + FN}$$

In [55]:
from sklearn.metrics import recall_score

In [56]:
# Modelo KNN
print(recall_score(y_teste, predito_knn))

0.46271186440677964


In [57]:
# Modelo Bernoulli de Naïve Bayes
print(recall_score(y_teste, predito_bnb))

0.8033898305084746


In [58]:
# Modelo Árvore de Decisão
print(recall_score(y_teste, predito_dtc))

0.5084745762711864


## Para saber mais: outras métricas

### F1 score

Além da acurácia, precisão e *recall*, podemos extrair da matriz de confusão uma outra métrica conhecida como **F1 score**, que é uma média harmônica entre o *recall* e a precisão. O *recall* e a precisão são grandezas inversamente proporcionais, ou seja, à medida que vai aumentando o *recall*, por consequência irá diminuir a precisão e vice-versa.

o *F1 score* é uma métrica para identificar se algum dos valores de *recall* ou precisão estão baixos, uma vez que não é muito interessante ter um modelo com um alto *recall* e pouca precisão, ou vice-versa. Então, caso sua intenção seja obter um valor equilibrado entre o *recall* e a precisão, a métrica que precisamos maximizar é o *F1 score*.

Para calcular, devemos utilizar a fórmula:

$$\Large F1 = 2 * \dfrac{precision * recall}{precision + recall}$$

### Curva ROC e AUC

Existe ainda uma métrica conhecida como AUC (Area Under the Curve), ou área sob a curva, que pode ser extraída da curva ROC. Essa métrica tem valor máximo de 1 e quanto maior o número mas bem avaliado será o modelo de classificação. Para a construção da curva ROC, precisamos utilizar diferentes pontos de corte para avaliação do modelo de classificação.

É possível criar diferentes matrizes de confusão a partir dos diversos pontos de corte e extrair as métricas de *recall* (taxa de verdadeiros positivos) e taxa de falsos positivos. Sendo esta última, o valor de falsos positivos divido pela soma de todos os valores negativos reais. Essas métricas podem ser sumarizadas em uma tabela e serão as coordenadas da curva ROC. Vamos entender como funcionam esses pontos de corte.

Na construção de um modelo de classificação, a predição de uma classe está atrelada a uma probabilidade. Imagine que temos uma variável alvo com duas possibilidades (0 ou 1). Se um modelo classifica uma observação como 1, significa que há uma probabilidade de x% de que aquela observação seja da classe 1.

O ponto de corte é um valor de probabilidade no qual, se a probabilidade de predição for maior que esse valor, atribui-se a observação à classe 1. Se for menor, atribui-se a observação à classe 0. À medida que o ponto de corte varia, os resultados obtidos na classificação se tornam diferentes, fazendo com que o modelo acerte mais de uma classe em troca de errar mais em outra.

Através de um exemplo, ficará evidente como os pontos de corte modificam as métricas da matriz de confusão. Na imagem abaixo estão representados a precisão e o *recall*. Nela estão definidos 3 pontos de corte em 25%, 50% e 75%, representados pelas setas verticais. Os zeros e uns são os valores reais das classes e a classificação será feita através dos pontos de corte. Os valores à esquerda da seta serão classificados como 0 e os valores à direita serão classificados como 1 pelo modelo.

![Diferentes pontos de corte](https://caelum-online-public.s3.amazonaws.com/2422-machine-learning/05/Aula5-img3.png)

No primeiro ponto de corte temos 3 valores cassificados como 0, à esquerda; e 9 valores como 1, à direita. A precisão é dada por 5/(5+4) = 55,5%, mostrando que 5 valores foram classificados corretamente como 1, dentre 9 valores totais classificados como 1. O *recall* é dado por 5/(5+1) = 83%, indicando que 5 valores foram classificados corretamente como 1, dentre 6 valores reais 1 no conjunto de dados.

No segundo ponto de corte temos 6 valores cassificados como 0 à esquerda; e 6 valores classificados como 1, à direita. A precisão é dada por 4/(4+2) = 66,6%, apontando que 4 valores foram classificados corretamente como 1, dentre 6 valores totais classificados como 1. O *recall* é dado por 4/(4+2) = 66,6%, demonstrando que 4 valores foram classificados corretamente como 1, dentre 6 valores reais 1 no conjunto de dados. O mesmo raciocínio pode ser aplicado para o último ponto de corte, resultando em valores diferentes para as métricas.

Para cada ponto de corte, uma matriz de confusão pode ser criada e as métricas podem ser extraídas em uma tabela, na qual cada linha representa um ponto de corte com suas respectivas métricas. Desta tabela, pode ser construído o gráfico da curva ROC. Esta curva mostra como o classificador se comporta para diferentes valores de pontos de corte, de acordo com a relação de verdadeiros positivos e falsos positivos.

![TP Rate & FP Rate](https://caelum-online-public.s3.amazonaws.com/2422-machine-learning/05/Aula5-img4.png)

Existe uma linha de referência na diagonal do gráfico que corresponde à uma linha de base e representa o caso no qual o classificador identifica aleatoriamente as classes. Mas como podemos interpretar esse gráfico?

Percebemos que a linha representando o classificador *RandomForestClassifier* cresce rapidamente até atingir o valor máximo de *True Positive Rate* (o *recall*). Isso significa que o modelo atinge 100% de classificação correta dos positivos e permanece até atingir o ponto (1,1) do gráfico. Esse ponto nos diz que o modelo tem 100% de *True Positive Rate* e 100% de *False Positive Rate* (taxa de falso negativo), indicando que ele classifica todas as amostras positivas corretamente e todas as amostras que não são positivas foram incorretamente classificadas.

Então, quando o modelo chega no 100% de *True Positive Rate* e lá permanece até atingir o 100% do *False Positive Rate*, ele consegue classificar todas as amostras positivas de forma correta independente do *threshold* adotado. Por outro lado, a linha representando o classificador *DecisionTreeClassifier* demora mais pra chegar no 100% de *True Positive Rate*. Porém, quando chega no valor máximo, permanece igual à do classificador anterior.

Esse pequeno atraso do *DecisionTreeClassifier* em relação ao *RandomForestClassifier* resulta em áreas diferentes, que são as AUCs para cada classificador. O primeiro teve uma AUC de 0.914 e o segundo 0.979. Portanto, esse último é o melhor classificador.